File: cpp/rpc/delegation/cache.cc
#include "rpc/delegation/cache.h"

namespace sigmaos {
namespace rpc {
namespace delegation {

bool Cache::_l = sigmaos::util::log::init_logger(RPCCLNT_CACHE);
bool Cache::_l_e = sigmaos::util::log::init_logger(RPCCLNT_CACHE_ERR);

void Cache::Register(uint64_t rpc_idx) {
  log(RPCCLNT_CACHE, "Register cachable DelegatedRPC RPC({})", (int)rpc_idx);
  std::lock_guard<std::mutex> lock(_mu);
  // Sanity check
  if (_registered.contains(rpc_idx)) {
    fatal("Double-register RPC idx {}", (int)rpc_idx);
  }
  _registered[rpc_idx] = true;
  _done[rpc_idx] = false;
}

void Cache::Put(uint64_t rpc_idx, std::shared_ptr<SigmaDelegatedRPCRep> rep,
                std::shared_ptr<sigmaos::serr::Error> err) {
  {
    log(RPCCLNT_CACHE, "Put cached DelegatedRPC RPC({}) err {}", (int)rpc_idx,
        (err != nullptr));
    std::lock_guard<std::mutex> lock(_mu);
    // Sanity check
    if (!_registered.contains(rpc_idx)) {
      fatal("Complete unregistered RPC({})", (int)rpc_idx);
    }
    // Sanity check
    if (_done.contains(rpc_idx) && _done.at(rpc_idx)) {
      fatal("Complete already-completed RPC({})", (int)rpc_idx);
    }
    // Store result
    _reps[rpc_idx] = rep;
    _done[rpc_idx] = true;
    _errors[rpc_idx] = err;
  }
  // After releasing the lock, broadcast to waiters that a reply has been
  // received
  _cond.notify_all();
}

std::expected<bool, sigmaos::serr::Error> Cache::Get(
    uint64_t rpc_idx, std::shared_ptr<sigmaos::io::iovec::IOVec> out_iov) {
  log(RPCCLNT_CACHE, "Get cached DelegatedRPC RPC({})", (int)rpc_idx);
  // Acquire the lock
  std::unique_lock<std::mutex> lock(_mu);

  // Delegated RPC retrieval is not in-progress, so bail out
  if (!_registered.contains(rpc_idx)) {
    log(RPCCLNT_CACHE, "No cached DelegatedRPC registered RPC({})",
        (int)rpc_idx);
    return false;
  }

  // Wait for the delegated RPC retrieval to complete
  _cond.wait(lock, [this, rpc_idx] { return _done.at(rpc_idx); });

  log(RPCCLNT_CACHE, "Done waiting for cached DelegatedRPC RPC({})",
      (int)rpc_idx);

  // Sanity check that the RPC indeed completed
  if (!_done.at(rpc_idx)) {
    fatal("Stopped waiting despite lack of RPC({}) completion", (int)rpc_idx);
  }
  // Copy reply to output
  auto cached_reply = _reps.at(rpc_idx);
  // Set the wrapper & serialized RPC buffers since these two are prepended
  // by the RPC stack and aren't supplied by the caller
  for (int i = 0; i < 2; i++) {
    out_iov->SetBuffer(i, std::make_shared<sigmaos::io::iovec::Buffer>(
                              cached_reply->mutable_blob()->mutable_iov(i)));
  }
  // Copy into the remaining IOVs in the blob, since these are supplied by the
  // caller
  for (int i = 2; i < cached_reply->blob().iov().size(); i++) {
    *(out_iov->GetBuffer(i)->Get()) = cached_reply->blob().iov(i);
  }
  //  *rep->mutable_err() = cached_reply->err();

  auto err = _errors.at(rpc_idx);
  if (err) {
    return std::unexpected(*_errors[rpc_idx]);
  }
  return true;
}

};  // namespace delegation
};  // namespace rpc
};  // namespace sigmaos

File: rpc/clnt/delegation/cache.go
package delegation

import (
	"sync"

	db "sigmaos/debug"
	spproxyproto "sigmaos/proxy/sigmap/proto"
)

// Client-side cache of delegated RPC replies
type ReplyCache struct {
	sync.Mutex
	cond       *sync.Cond
	registered map[uint64]bool
	done       map[uint64]bool
	reps       map[uint64]*spproxyproto.SigmaDelegatedRPCRep
	errors     map[uint64]error
}

func NewReplyCache() *ReplyCache {
	c := &ReplyCache{
		registered: make(map[uint64]bool),
		done:       make(map[uint64]bool),
		reps:       make(map[uint64]*spproxyproto.SigmaDelegatedRPCRep),
		errors:     make(map[uint64]error),
	}
	c.cond = sync.NewCond(&c.Mutex)
	return c
}

// Register an in-progress delegated RPC retrieval
func (c *ReplyCache) Register(idx uint64) {
	c.Lock()
	defer c.Unlock()

	// Sanity check
	if c.registered[idx] {
		db.DFatalf("Double-register delegated RPC(%v)", idx)
	}
	c.registered[idx] = true
	c.done[idx] = false
}

func (c *ReplyCache) Get(idx uint64, rep *spproxyproto.SigmaDelegatedRPCRep) (error, bool) {
	c.Lock()
	defer c.Unlock()
	outiov := rep.Blob.Iov

	// Delegated RPC retrieval is not in-progress, so bail out
	if !c.registered[idx] {
		return nil, false
	}

	// Wait for the delegated RPC retrieval to complete
	for !c.done[idx] {
		c.cond.Wait()
	}
	// Copy reply to outiov
	cachedReply := c.reps[idx]
	// First two entries are the serialized RPC wrappers
	outiov[0] = cachedReply.Blob.Iov[0]
	outiov[1] = cachedReply.Blob.Iov[1]
	// Remaining entries are user-supplied buffers in the RPC Blob, so we need to
	// copy to them
	for i := 2; i < len(outiov); i++ {
		copy(outiov[i], cachedReply.Blob.Iov[i])
	}
	rep.Err = cachedReply.Err
	// Return the result
	return c.errors[idx], true
}

func (c *ReplyCache) Put(idx uint64, rep *spproxyproto.SigmaDelegatedRPCRep, err error) {
	c.Lock()
	defer c.Unlock()

	// Sanity check
	if !c.registered[idx] {
		db.DFatalf("Complete unregistered RPC(%v)", idx)
	}
	// Sanity check
	if c.done[idx] {
		db.DFatalf("Complete already-completed RPC(%v)", idx)
	}
	// Store result
	c.reps[idx] = rep
	c.errors[idx] = err
	c.done[idx] = true
	// Wake up waiters
	c.cond.Broadcast()
}
